lib/active_merchant/billing/gateways/migs.rb

```diff
@@ -173,6 +177,7 @@ def purchase_offsite_url(money, options = {})
         post[:ReturnURL] = options[:return_url]
 
         add_standard_parameters('pay', post, options[:unique_id])
+        add_tx_source(post, options)
```

Comment:
It looks like we don't have a test for this method (```purchase_offsite_url```). The test to add would just be that ```tx_source``` is added to the URL, which the method returns.

Once a test is added, you can print out the return value from ```purchase_offsite_url``` to see what the result should look like, and test that ```vpc_TxSource``` is present in that string. The return value will look something like this:

```
https://migs.mastercard.com.au/vpcpay?vpc_Amount=100&vpc_OrderInfo=1&vpc_Locale=en&vpc_ReturnURL=http%3A%2F%2Fwww.aol.com%2F&vpc_Version=1&vpc_Merchant=login&vpc_AccessCode=password&vpc_Command=pay&vpc_MerchTxnRef=fe9f392f1ea3e2fba612dd176acfd7ee&vpc_TxSource=MOTO&vpc_SecureHash=06355CAF2AB6AA8754583456F400926FAA1911EA4575E4743E85C8DA66FF4B71&vpc_SecureHashType=SHA256
```

Reply:
Do we know whether the `purchase_offsite_url` actually needs the `tx_source` parameter passed? I added it in, but I'm still not sure whether it should be there.

Reply:
Currently, my remote test only checks that the `purchase` action passes `tx_source`. Since I'm adding a unit test to ensure that every method can pass `tx_source`, should I also have remote tests for each method?

Reply:
Based on the code comments, ```purchase_offsite_url``` appears to be used for a payment flow where the user clicks a link that goes to MiGS for the payment processing. It makes sense that all parameters for the transaction would have to appear in that URL (where they are query parameters). So, I think ```purchase_offsite_url``` will need to include the options in ```purchase``` (except for the credit card itself).

The remote tests call out to gateways to test how they respond to our requests. ```purchase_offsite_url``` just generates a URL string with no external communication, so a unit test is all we need there.

Reply:
After digging slightly deeper, I see that there is an existing remote test that simulates clicking on the URL that ```purchase_offsite_url``` generates: https://github.com/activemerchant/active_merchant/blob/547d0ebb419f2070f8327ae30d1b0dc7187c7b94/test/remote/gateways/remote_migs_test.rb#L34. This is pretty cool. We could use that test as the basis for an additional one that tests the offsite URL with the ```vpc_TxSource``` option. The assertion would likely be the same as the existing test (i.e., we just need to confirm that a successful response still comes back when the new option is present).

Reply:
The `purchase_offsite_url` method, in addition to using `add_creditcard_type` instead of the `add_creditcard` method used by the `purchase` action, does not include the `add_3ds` method, so it's already not receiving all of the parameters that the `purchase` action does. 

Since the `purchase_offsite_url` method states that it "Generates a URL to redirect user to MiGS to process payment", it seems as though the source of the transaction would be readily apparent and potentially would not need to be specified. Is there a way to determine whether this method requires `tx_source` to be passed?

Reply:
> Currently, my remote test only checks that the `purchase` action passes `tx_source`.

My question here was not specifically in regards to the `purchase_offsite_url` method. Should I add separate remote tests for all of the core actions to which I added `add_tx_source`, i.e., `purchase`, `capture`, `refund`, and `void`?

Reply:
Yep -- testing each of the calls against the real gateway would be good to confirm that they behave as expected. For the offsite URL, I'd expect the documentation to have this information (somewhere). If there's still uncertainty about it, I'd see including it as low risk assuming the remote test verifies that the gateway doesn't produce an error when it is there. The code as you wrote it only will include the parameter if the user has specified it, and the user  would likely only include it if is needed, so that makes it even less of a risk if we're not able to pinpoint that in the docs.

Reply:
@jasonxp The `test_server_purchase_url` method, which you posted above and suggested for use as the basis of a test for the `purchase_offsite_url` action passing `tx_source`, is the method that is returning the error that I've been overlooking when running the full remote test file. Since that method is failing, can you think of another way to write a test for `purchase_offsite_url` passing `tx_source`?

Reply:
In that case: @curiousepic mentioned that this particular flow isn't supported by Spreedly. I'd thus be inclined to not bother adding the option to ```purchase_offsite_url```, since the request to add ```vpc_TxSource``` originated with a Spreedly customer, and the additional testing would be inconvenient. Would you agree, @curiousepic?

Reply:
I think that's the better of two suboptimal situations, yep 👍

Reply:
I've submitted a new commit that removes `add_tx_source` from the `purchase_offsite_url` method.